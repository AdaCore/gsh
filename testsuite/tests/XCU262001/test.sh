echo "Operators - = ? and +"
echo "===================="

echo ""
echo "expansion with -"
param="abcd"
empty_str=""
echo '${param-myword}:     ' ${param-myword} '     / expected abcd'
echo '${empty_str-myword}: ' ${empty_str-myword} '          / expected null'
unset unset_var
echo '${unset_var-myword}: ' ${unset_var-myword} '   / expected myword'

echo ""
echo "expansion with ="
param="abcd"
empty_str=""
echo '${param=myword}:     ' ${param=myword} '      / expected abcd'
echo '${empty_str=myword}: ' ${empty_str=myword} '           / expected null'
unset unset_var
echo '${unset_var=myword}: ' ${unset_var=myword} '    / expected unset_var is assigned myword:' $unset_var

echo ""
echo "expansion with ?"
param="abcd"
empty_str=""
echo '${param?myword}:      ' ${param?myword} '     / expected abcd'
echo '${empty_str?myword}:  ' ${empty_str?myword} '          / expected null'
unset unset_var
(
echo '${unset_var?myword}:  ' ${unset_var?myword} '     / expected error'
)

echo ""
echo "expansion with +"
param="abcd"
empty_str=""
echo '${param+myword}:      ' ${param+myword} '   / expected myword'
echo '${empty_str+myword}:  ' ${empty_str+myword} '   / expected myword'
unset unset_var
echo '${unset_var+myword}:  ' ${unset_var+myword} '          / expected null'

echo ""
echo "Operators - = ? and + introduced by COLON ':'"
echo "=============================================="
echo ""
echo "expansion with :-"
param="abcd"
empty_str=""
echo '${param:-myword}:      ' ${param:-myword} '     / expected abcd'
echo '${empty_str:-myword}:  ' ${empty_str:-myword} '   / expected myword'
unset unset_var
echo '${unset_var:-myword}:  ' ${unset_var:-myword} '   / expected myword'

echo ""
echo "expansion with :="
param="abcd"
empty_str=""
echo '${param:=myword}:      ' ${param:=myword} '     / expected abcd'
echo '${empty_str:=myword}:  ' ${empty_str:=myword} '   / expected empty_str is assigned myword:' $empty_str
unset unset_var
echo '${unset_var:=myword}:  ' ${unset_var:=myword} '   / expected unset_var is assigned myword:' $unset_var

echo ""
echo "expansion with :?"
param="abcd"
empty_str=""
echo '${param:?myword}:     ' ${param:?myword} '     / expected abcd'
(
echo '${empty_str:?myword}: ' ${empty_str:?myword} / expected error
)
(
unset unset_var
echo '${unset_var?myword}:  ' ${unset_var:?myword} / expected error
)

echo ""
echo "expansion with :+"
param="abcd"
empty_str=""
echo '${param:+myword}:     ' ${param:+myword}  '   / expected myword'
echo '${empty_str:+myword}: ' ${empty_str:+myword} '          / expected null'
unset unset_var
echo '${unset_var:+myword}: ' ${unset_var:+myword} '          / expected null'


echo ""
echo "Length (subst starting with by #)"
echo "================================="
param="abcd"
empty_str=""
unset unset_var
echo '${#param}:     ' ${#param} / expected 4
echo '${#empty_str}: ' ${#empty_str} / expected 0
echo '${#unset_var}: '  ${#unset_var} / expected 0


echo ""
echo "Substrings processing"
echo "====================="

echo ""
echo "expansion with %"
x=file.c
empty_str=""
echo currently x is $x
echo '${x%.c}.o:         ' ${x%.c}.o '    / expected file.o'
echo '${empty_str%.c}:   ' ${empty_str%.c} '           / expected null'
unset unset_var
echo '${unset_var%abc}:  ' ${unset_var%abc} '           / expected null'

echo ""
echo "expansion with %%"
x=posix/src/std
empty_str=""
echo currently x is $x
echo '${x%%/*}:                     ' ${x%%/*} '           / expected posix'
x=posix/src/std
echo currently x is $x
echo '${x%%*/}:                     ' ${x%%*/} '   / expected posix/src/std'
echo '${empty_str%%/*}:             ' ${empty_str%%/*} '                 / expected null'
unset unset_var
echo '${unset_var%%abc}:            ' ${unset_var%%abc} '                 / expected null'

echo ""
echo "expansion with #"
x=$HOME/src/cmd
empty_str=""
echo currently x is '$HOME/src/cmd'
echo '${x#$HOME}:                  ' ${x#$HOME} '   / expected /src/cmd'
x=posix/src/std
echo currently x is $x
echo '${x#*/}:                     ' ${x#*/} '   / expected src/std'
echo '${empty_str#abc}:            ' ${empty_str#abc} '           / expected null'
unset unset_var
echo '${unset_var#abc}:            ' ${unset_var#abc} '           / expected null'

echo ""
echo "expansion with ##"
x=/one/two/three
empty_str=""
echo currently x is $x
echo '${x##*/}:          ' ${x##*/} '   / expected three'
echo '${empty_str##abc}: ' ${empty_str##abc} '         / expected null'
unset unset_var
echo '${unset_var##abc}: ' ${unset_var##abc} '         / expected null'

