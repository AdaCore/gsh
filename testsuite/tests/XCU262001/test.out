Operators - = ? and +
====================

expansion with -
${param-myword}:      abcd      / expected abcd
${empty_str-myword}:            / expected null
${unset_var-myword}:  myword    / expected myword

expansion with =
${param=myword}:      abcd       / expected abcd
${empty_str=myword}:             / expected null
${unset_var=myword}:  myword     / expected unset_var is assigned myword: myword

expansion with ?
${param?myword}:       abcd      / expected abcd
${empty_str?myword}:             / expected null
[error] unset_var: myword

expansion with +
${param+myword}:       myword    / expected myword
${empty_str+myword}:   myword    / expected myword
${unset_var+myword}:             / expected null

Operators - = ? and + introduced by COLON ':'
==============================================

expansion with :-
${param:-myword}:       abcd      / expected abcd
${empty_str:-myword}:   myword    / expected myword
${unset_var:-myword}:   myword    / expected myword

expansion with :=
${param:=myword}:       abcd      / expected abcd
${empty_str:=myword}:   myword    / expected empty_str is assigned myword: myword
${unset_var:=myword}:   myword    / expected unset_var is assigned myword: myword

expansion with :?
${param:?myword}:      abcd      / expected abcd
[error] empty_str: myword
[error] unset_var: myword

expansion with :+
${param:+myword}:      myword    / expected myword
${empty_str:+myword}:            / expected null
${unset_var:+myword}:            / expected null

Length (subst starting with by #)
=================================
${#param}:      4 / expected 4
${#empty_str}:  0 / expected 0
${#unset_var}:  0 / expected 0

Substrings processing
=====================

expansion with %
currently x is file.c
${x%.c}.o:          file.o     / expected file.o
${empty_str%.c}:               / expected null
${unset_var%abc}:              / expected null

expansion with %%
currently x is posix/src/std
${x%%/*}:                      posix            / expected posix
currently x is posix/src/std
${x%%*/}:                      posix/src/std    / expected posix/src/std
${empty_str%%/*}:                               / expected null
${unset_var%%abc}:                              / expected null

expansion with #
currently x is $HOME/src/cmd
${x#$HOME}:                   /src/cmd    / expected /src/cmd
currently x is posix/src/std
${x#*/}:                      src/std    / expected src/std
${empty_str#abc}:                        / expected null
${unset_var#abc}:                        / expected null

expansion with ##
currently x is /one/two/three
${x##*/}:           three    / expected three
${empty_str##abc}:           / expected null
${unset_var##abc}:           / expected null
